share/search/daemon.py

```diff
@@ -0,0 +1,104 @@
+
+from bots.elasticsearch.tasks import index_model
+
+
+logger = logging.getLogger(__name__)
+
+
+class SearchIndexerDaemon:
+
+    def __init__(self, celery_app, max_size=500, timeout=5, flush_interval=10):
+        self.app = celery_app
+        self.buffer = []
+        self.messages = []
+        self.last_flush = 0
+
+        self.flush_interval = flush_interval
+        self.max_size = max_size
+        self.timeout = timeout
+
+    def run(self):
+        try:
+            connection = self.app.pool.acquire(block=True)
+            queue = connection.SimpleQueue(settings.ELASTIC_QUEUE, **settings.ELASTIC_QUEUE_SETTINGS)
+        except Exception as e:
+            logger.exception('Could not connect to broker')
+            raise
+
+        logger.info('Connected to broker')
+        logger.info('Using queue "%s"', settings.ELASTIC_QUEUE)
+
+        try:
+            self._run(queue)
+        except KeyboardInterrupt:
+            logger.warning('Recieved Interrupt. Exiting...')
+            return
+        except Exception as e:
+            logger.exception('Encountered an unexpected error. Attempting to flush before exiting.')
+            if self.buffer or self.messages:
+                try:
+                    self.flush()
+                except Exception:
+                    logger.exception('%d messages could not be flushed', len(self.messages))
+            raise e
+        finally:
+            try:
+                queue.close()
+                connection.close()
+            except Exception as e:
+                logger.exception('Failed to clean up broker connection')
+
+    def flush(self):
+        logger.info('Flushing %d messages', len(self.messages))
+
+        # TODO Move search logic into this module
+        # TODO Support more than just creativeworks?
+        logger.debug('Sending %d works to Elasticsearch', len(self.buffer))
+        index_model('CreativeWork', self.buffer)
+        logger.debug('Works successfully indexed')
+
+        errors = []
+        logger.debug('ACKing received messages')
+        for message in self.messages:
+            try:
+                message.ack()
+            except Exception as e:
+                logger.exception('Could not ACK %r', message)
+                errors.append(e)
+        logger.debug('Messages successfully ACKed')
+
+        self.buffer.clear()
+        self.messages.clear()
+        self.last_flush = time.time()
+
+        if errors:
+            logger.error('Encounted %d errors while attempting to ACK messages', len(errors))
+            raise errors[0]
+        logger.debug('Successfully flushed messages')
+
+    def _run(self, queue):
+        self.last_flush = time.time()
+
+        while True:
+            try:
+                message = queue.get(timeout=self.timeout)
+                self.messages.append(message)
+                # TODO Support more than just creativeworks?
+                self.buffer.extend(message.payload.get('CreativeWork', []))
```

Comment:
Why keep a separate buffer of works, rather than extracting the works in `flush()`? More buffers more problems.

Would there be a downside if `max_size` limited `messages` instead?